### PR TITLE
attempt to fix #114, but it's not working...

### LINKED ISSUE #114: CUDA ERROR an illegal memory access was encountered
commit [`c02bae2ebc684a2e068c0dc59be00ff43167b44d`
](https://github.com/karpathy/llm.c/commit/c02bae2ebc684a2e068c0dc59be00ff43167b44d)
I got this error when running `train_gpt2cu`
```
[System]
Device 0: NVIDIA GeForce RTX 3060
enable_tf32: 1
[GPT-2]
max_seq_len: 1024
vocab_size: 50257
num_layers: 12
num_heads: 12
channels: 768
num_parameters: 124439808
train dataset num_batches: 74
val dataset num_batches: 8
batch size: 4
sequence length: 1024
val_num_batches: 10
num_activations: 2456637440
val loss 4.517170
step 0: train loss 4.367660 (took 140.356816 ms)
step 1: train loss 4.406325 (took 140.393840 ms)
step 2: train loss 4.484698 (took 140.358579 ms)
step 3: train loss 4.345248 (took 140.415392 ms)
step 4: train loss 4.043498 (took 140.366040 ms)
step 5: train loss 4.229475 (took 140.349640 ms)
step 6: train loss 4.175006 (took 140.337543 ms)
step 7: train loss 4.207677 (took 140.347408 ms)
step 8: train loss 4.127449 (took 140.839560 ms)
step 9: train loss 4.220684 (took 141.022061 ms)
val loss 4.517170
step 10: train loss 4.345275 (took 140.932235 ms)
step 11: train loss 4.245928 (took 141.035935 ms)
step 12: train loss 4.160665 (took 140.851916 ms)
step 13: train loss 3.989787 (took 140.892920 ms)
step 14: train loss 4.306040 (took 140.957500 ms)
step 15: train loss 4.340417 (took 141.032707 ms)
step 16: train loss 4.304399 (took 141.010662 ms)
step 17: train loss 4.424002 (took 140.930450 ms)
step 18: train loss 4.314507 (took 141.028123 ms)
step 19: train loss 4.287285 (took 140.992381 ms)
val loss 4.517170
[CUDA ERROR] at file train_gpt2.cu:1211:
an illegal memory access was encountered
```

### `nvidia-smi`
```
(.venv) ➜  llm.c git:(dev) ✗ nvidia-smi
Sun Apr 14 05:29:02 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3060        Off |   00000000:01:00.0 Off |                  N/A |
|  0%   32C    P8             13W /  170W |       3MiB /  12288MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+
```

### `nvcc`
```
(.venv) ➜  llm.c git:(dev) ✗ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Thu_Mar_28_02:18:24_PDT_2024
Cuda compilation tools, release 12.4, V12.4.131
Build cuda_12.4.r12.4/compiler.34097967_0
```

The file `train_gpt2.cu:1211:` indicates:
```C
cudaCheck(cudaMemcpy(cpu_probs, probs, model.config.vocab_size * sizeof(float), cudaMemcpyDeviceToHost));
```

### PR DESCRIPTION
**Pull request by karpathy**
so to fix #114 :
I tried to do the inference keeping the batch always the same size B,T

but this is not resolving the issue.
unfortunately I can't spare time to keep looking at this tonight (until maybe very late), so I'm keeping this unresolved for now. 

I suspect there is something wrong with kernel5, we did not check it properly when we merged, blinded by the speed it offered sigh.

### CODE REVIEW

**Comment by karpathy** [comment]
If there is a bug with kernel5, I don't really understand how we're lucking out that test_gpt2.cu reports no issues, and the logits and the loss (the output of the forward pass) match expected values.

**Comment by yuanrongxi** [comment]
I attempted to test according to the commits in this PR, but I still encountered 'an illegal memory access was encountered.'
[System]
Device 0: NVIDIA GeForce GTX 1650 Ti
enable_tf32: 0
[GPT-2]
max_seq_len: 1024
vocab_size: 50257
num_layers: 12
num_heads: 12
channels: 768
num_parameters: 124439808
train dataset num_batches: 74
val dataset num_batches: 8
batch size: 4
sequence length: 1024
val_num_batches: 10
num_activations: 2456637440
val loss 4.517293
step 0: train loss 4.367856 (took 12696.786693 ms)
step 1: train loss 4.406481 (took 12592.453470 ms)
step 2: train loss 4.484837 (took 12581.640492 ms)
step 3: train loss 4.345325 (took 12785.257593 ms)
step 4: train loss 4.043289 (took 12699.167329 ms)
step 5: train loss 4.229304 (took 12977.969745 ms)
step 6: train loss 4.174998 (took 13413.872346 ms)
step 7: train loss 4.207466 (took 13131.313881 ms)
step 8: train loss 4.127153 (took 13543.845417 ms)
step 9: train loss 4.220581 (took 13296.718636 ms)
val loss 4.517293
step 10: train loss 4.345143 (took 13550.681370 ms)
step 11: train loss 4.245717 (took 13489.289869 ms)
step 12: train loss 4.160386 (took 13409.223163 ms)
step 13: train loss 3.989360 (took 13403.222145 ms)
step 14: train loss 4.305948 (took 13426.626539 ms)
step 15: train loss 4.340422 (took 13489.533329 ms)
step 16: train loss 4.304480 (took 13612.480865 ms)
step 17: train loss 4.424025 (took 13379.971251 ms)
step 18: train loss 4.314513 (took 13098.182640 ms)
step 19: train loss 4.287313 (took 13259.752212 ms)
val loss 4.517293
[CUDA ERROR] at file train_gpt2.cu:1241:
an illegal memory access was encountered

**Comment by ngc92** [comment]
@karpathy 
OK, so if I comment out enough stuff, so that we don't run the validation part, and immediately jump into generation with a fixed T=64, compute-sanitizer actually does report some error:
```
========= Invalid __global__ read of size 4 bytes
=========     at encoder_forward_kernel2(float *, int *, float *, float *, int, int, int)+0x410 in train_gpt2.cu:89
=========     by thread (0,0,0) in block (930,0,0)


========= Invalid __global__ read of size 4 bytes
=========     at encoder_forward_kernel2(float *, int *, float *, float *, int, int, int)+0x410 in train_gpt2.cu:89
=========     by thread (1,0,0) in block (930,0,0)
=========     Address 0x7f4676000004 is out of bounds
=========     and is 522 190 853 bytes after the nearest allocation at 0x7f4656600000 of size 8 388 608 bytes
````

**Comment by ngc92** [comment]
@karpathy 
I think it's a padding problem, see #122
